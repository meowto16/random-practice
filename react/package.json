{
  "name": "react-experiments",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@hookform/resolvers": "^2.9.8",
    "@vkontakte/icons": "^1.192.0",
    "@vkontakte/vk-bridge": "^2.7.0",
    "@vkontakte/vkui": "^4.37.2",
    "axios": "^0.27.2",
    "nock": "^13.2.9",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-hook-form": "^7.35.0",
    "react-scripts": "5.0.1",
    "web-vitals": "^2.1.4",
    "yup": "^0.32.11"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "devDependencies": {
    "@faker-js/faker": "^7.5.0",
    "@testing-library/jest-dom": "^5.16.5",
    "@testing-library/react": "^13.4.0",
    "@testing-library/user-event": "^13.5.0"
  }
}
