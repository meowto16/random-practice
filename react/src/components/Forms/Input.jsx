import { Input as VKInput } from '@vkontakte/vkui'

import { FormItem } from './FormItem'

export const Input = ({
  FormItemProps,
  InputProps,
  name,
  label,
  placeholder,
  type,
}) => (
  <FormItem name={name} label={label} {...FormItemProps}>
    <VKInput
      {...InputProps}
      aria-label={label}
      type={type}
      placeholder={placeholder}
    />
  </FormItem>
)
