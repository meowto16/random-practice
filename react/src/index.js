import { ConfigProvider } from '@vkontakte/vkui'

import '@vkontakte/vkui/dist/vkui.css'
import React from 'react'
import ReactDOM from 'react-dom/client'

import App from './App'
import './index.css'

const root = ReactDOM.createRoot(document.getElementById('root'));
root.render(
  <React.StrictMode>
    <ConfigProvider appearance="dark" isWebView={false} platform="ios">
      <App />
    </ConfigProvider>
  </React.StrictMode>
);
