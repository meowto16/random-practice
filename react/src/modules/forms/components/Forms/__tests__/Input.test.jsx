import { act, fireEvent } from '@testing-library/react'
import * as yup from 'yup'

import { Input } from '../Input'

import { renderWithReactHookForm } from '../../../../utils.testing'

const eventMock = expect.anything();

describe('Input (Form component)', () => {
  it('renders without errors', () => {
    const { screen } = renderWithReactHookForm(<Input name="firstname" label="Имя" />);
    screen.unmount();
  })

  it('same as snapshot', () => {
    const { screen } = renderWithReactHookForm(<Input name="firstname" label="Имя" />);
    expect(screen.container.firstChild).toMatchSnapshot();
  })

  it('supports default value', () => {
    const { screen } = renderWithReactHookForm(
      <Input name="firstname" label="Введите имя" />,
      {
        defaultValues: {
          firstname: 'Максим'
        }
    });

    expect(screen.getByDisplayValue('Максим')).toBeInTheDocument()
  })

  it('changes form value (react-hook-form context)', async () => {
    const { screen, form, submitHandler } = renderWithReactHookForm(<Input name="firstname" label="Имя" placeholder="Введите имя" />)

    // Check empty value
    await act(() => {
      fireEvent.submit(form)
    })
    expect(submitHandler).toHaveBeenCalledWith({ firstname: '' }, eventMock)

    // Change value
    const input = screen.getByPlaceholderText('Введите имя')
    await act(() => {
      fireEvent.input(input, { target: { value: 'Максим' } })
      fireEvent.submit(form)
    })

    // Check changed value
    expect(submitHandler).toHaveBeenCalledWith({ firstname: 'Максим' }, eventMock)
  })

  it('supports validation', async () => {
    const { screen, form } = renderWithReactHookForm(
      <Input name="firstname" label="Имя" placeholder="Введите имя" />,
      {
        validationSchema: yup.object({
          firstname: yup.string().required('Это обязательное поле')
        }),
      }
    )

    // Error is not visible before submit form
    expect(screen.queryByText('Это обязательное поле')).not.toBeTruthy()

    // Submit form
    await act(() => void fireEvent.submit(form))

    // Error is now visible after submit form
    expect(screen.queryByText('Это обязательное поле')).toBeTruthy()
  })
})