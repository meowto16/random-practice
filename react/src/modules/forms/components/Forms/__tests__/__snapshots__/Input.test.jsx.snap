// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input (Form component) same as snapshot 1`] = `
<form
  name="form"
>
  <div
    class="vkuiFormItem vkuiFormItem--android FormItem FormItem--android vkuiFormItem--default FormItem--default vkuiFormItem--sizeY-regular FormItem--sizeY-regular vkuiFormItem--withTop FormItem--withTop"
  >
    <h5
      class="vkuiFormItem__top FormItem__top vkuiSubhead Subhead vkuiSubhead--sizeY-regular Subhead--sizeY-regular"
    >
      Имя
    </h5>
    <div
      class="vkuiInput Input vkuiInput--sizeY-regular Input--sizeY-regular vkuiFormField FormField vkuiFormField--default FormField--default vkuiFormField--status-default FormField--status-default vkuiFormField--sizeY-regular FormField--sizeY-regular"
      role="presentation"
    >
      <input
        aria-label="Имя"
        class="vkuiInput__el Input__el"
        label="Имя"
        name="firstname"
        type="text"
      />
      <div
        class="vkuiFormField__border FormField__border"
        role="presentation"
      />
    </div>
  </div>
  <button
    type="submit"
  >
    Отправить
  </button>
</form>
`;
